gibbon-compiler/src/Gibbon/Passes/HoistExpressions.hs

```diff
@@ -0,0 +1,340 @@
+        LetRegE reg rhs bod -> do
+          (bod', env') <- collectBoundsCheckExprs env benv bod
+          return (Ext $ LetRegE reg rhs bod', env')
+        RetE {} -> return (ex, env)
+        TagCursor {} -> return (ex, env)
+        LetRegionE r sz ty bod -> do
+          (bod', env') <- collectBoundsCheckExprs env benv bod
+          return (Ext $ LetRegionE r sz ty bod', env')
+        LetParRegionE r sz ty bod -> do
+          (bod', env') <- collectBoundsCheckExprs env benv bod
+          return (Ext $ LetParRegionE r sz ty bod', env')
+        FromEndE {} -> return (ex, env)
+        BoundsCheck {} -> return (ex, env)
+        IndirectionE {} -> return (ex, env)
+        GetCilkWorkerNum -> return (ex, env)
+        LetAvail vs bod -> do
+          (bod', env') <- collectBoundsCheckExprs env benv bod
+          return (Ext $ LetAvail vs bod', env')
+        AllocateTagHere {} -> return (ex, env)
+        AllocateScalarsHere {} -> pure (ex, env)
+        SSPush {} -> pure (ex, env)
+        SSPop {} -> pure (ex, env)
+        _ -> pure (ex, env)
+    -- Straightforward recursion
+    VarE {} -> return (ex, env)
+    LitE {} -> return (ex, env)
+    CharE {} -> return (ex, env)
+    FloatE {} -> return (ex, env)
+    LitSymE {} -> return (ex, env)
+    PrimAppE {} -> return (ex, env)
+    DataConE {} -> return (ex, env)
+    ProjE i e -> do
+      (e', env') <- collectBoundsCheckExprs env benv e
+      return (ProjE i e', env')
+    IfE a b c -> do
+      (a', env') <- collectBoundsCheckExprs env benv a
+      (b', env'') <- collectBoundsCheckExprs env benv b
+      (c', env''') <- collectBoundsCheckExprs env benv c
+      let env'''' = mergeHoistExprMaps [env', env'', env''']
+      return (IfE a' b' c', env'''')
+    MkProdE ls -> do
+      res <- mapM (collectBoundsCheckExprs env benv) ls
+      let ls' = map fst res
+      let env' = mergeHoistExprMaps $ map snd res
+      return (MkProdE ls', env')
+    CaseE scrt mp -> do
+      res <-
+        mapM
+          ( \(c, args, ae) -> do
+              (ae', env') <- collectBoundsCheckExprs env benv ae
+              return ((c, args, ae'), env')
+          )
+          mp
+      let mp' = map fst res
+      let env' = mergeHoistExprMaps $ map snd res
+      return (CaseE scrt mp', env')
+    TimeIt e ty b -> do
+      (e', env') <- collectBoundsCheckExprs env benv e
+      return (TimeIt e' ty b, env')
+    SpawnE {} -> error "collectBoundsCheckExprs: TODO SpawnE"
+    SyncE -> pure (ex, env)
+    MapE {} -> error $ "collectBoundsCheckExprs: TODO MapE"
+    FoldE {} -> error $ "collectBoundsCheckExprs: TODO FoldE"
+
+storeHoistableExpr :: FreeVarsTy -> FreeVarsTy -> S.Set FreeVarsTy -> HoistableExpr -> HoistAbleExprMap -> (HoistAbleExprMap, Bool)
+storeHoistableExpr v1 v2 dependentVars b env =
+  if v1 == v2
+    then
+      let delayLetBind = b
+          env' = M.insert delayLetBind dependentVars env
+       in (env', True)
+    else (env, False)
```

Comment:
Can you rename the function parameter from `b` to `delayLetBind` (is it `delayedLetBind` rather?) and avoid this let? Also, env' isn't necessary. Also, guards are preferable to if/then/else in Haskell.
```suggestion
storeHoistableExpr v1 v2 dependentVars delayedLetBind env =
  | v1 == v2 = 
      (M.insert delayLetBind dependentVars env, True)
  | otherwise = (env, False)
```